packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts

```diff
@@ -0,0 +1,148 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Node, RelationField, RelationshipSubscriptionsEvent, SubscriptionsEvent } from "../../../../../types";
+import type { ObjectFields } from "../../../../get-obj-field-meta";
+import type { RecordType, RelationshipType, SubscriptionContext } from "../../types";
+import { filterByProperties } from "./filter-by-properties";
+import { multipleConditionsAggregationMap } from "../utils/multiple-conditions-aggregation-map";
+import { filterRelationshipKey } from "../utils/filter-relationship-key";
+import { filterByValues } from "../../../../../translate/authorization/utils/filter-by-values";
+import type { SubscriptionsAuthorizationWhere } from "../../../../../schema-model/annotation/SubscriptionsAuthorizationAnnotation";
+
+function isRelationshipSubscriptionsEvent(event: SubscriptionsEvent): event is RelationshipSubscriptionsEvent {
+    return ["create_relationship", "delete_relationship"].includes(event.event);
+}
+
+export function filterByAuthorizationRules({
+    node,
+    where,
+    event,
+    nodes,
+    relationshipFields,
+    context,
+}: {
+    node: Node;
+    where:
+        | SubscriptionsAuthorizationWhere
+        | Record<
+              string,
+              RecordType | Record<string, RecordType | RelationshipType> | Array<Record<string, RecordType>>
+          >;
+    event: SubscriptionsEvent;
+    nodes?: Node[];
+    relationshipFields?: Map<string, ObjectFields>;
+    context: SubscriptionContext;
+}): boolean {
+    const receivedEventProperties = event.properties;
+
+    const results = Object.entries(where).map(([wherePropertyKey, wherePropertyValue]) => {
+        if (Object.keys(multipleConditionsAggregationMap).includes(wherePropertyKey)) {
+            const comparisonResultsAggregationFn = multipleConditionsAggregationMap[wherePropertyKey];
+            let comparisonResults;
+            if (wherePropertyKey === "NOT") {
+                comparisonResults = filterByAuthorizationRules({
+                    node,
+                    where: wherePropertyValue as Record<string, RecordType>,
+                    event,
+                    nodes,
+                    relationshipFields,
+                    context,
+                });
+            } else {
+                comparisonResults = (wherePropertyValue as Array<Record<string, RecordType>>).map((whereCl) => {
+                    return filterByAuthorizationRules({
+                        node,
+                        where: whereCl,
+                        event,
+                        nodes,
+                        relationshipFields,
+                        context,
+                    });
+                });
+            }
+
+            if (!comparisonResultsAggregationFn(comparisonResults)) {
+                return false;
+            }
```

Comment:
This code block is missing a lot of inferences from Typescript, resulting in many `any`, which makes `comparisonResults` here really hard to reason about. I can't tell what type it is, or what type comparisonResultsAggregationFn is.

Reply:
Honestly, I was really struggling with this code when I wrote it. It inherited from the existing Subscriptions events filtering, but then added _more_. The typings were crazy as they were, so adding more functionality was a nightmare. Happy to accept a hand through this if you want to have a crack at it!